syntax = "proto3";

package proto;

option go_package = "github.com/Jorrit05/DYNAMOS/pkg/proto";
import "google/protobuf/empty.proto";
import "google/protobuf/any.proto";
import "generic.proto";
import "microserviceCommunication.proto";

// The sidecar definition.
service SideCar {
  rpc InitRabbitMq(InitRequest) returns (google.protobuf.Empty) {}
  rpc Consume(ConsumeRequest) returns (stream SideCarMessage) {}
  rpc ChainConsume(ConsumeRequest) returns (stream SideCarMessage) {}
  rpc SendRequestApproval(RequestApproval) returns  (google.protobuf.Empty) {}
  rpc SendValidationResponse(ValidationResponse) returns  (google.protobuf.Empty) {}
  rpc SendCompositionRequest(CompositionRequest) returns  (google.protobuf.Empty) {}
  rpc SendSqlDataRequest(SqlDataRequest) returns  (google.protobuf.Empty) {}
  rpc SendPolicyUpdate(PolicyUpdate) returns  (google.protobuf.Empty) {}
  rpc SendTest(SqlDataRequest) returns  (google.protobuf.Empty) {}
  rpc SendMicroserviceComm(MicroserviceCommunication) returns  (google.protobuf.Empty) {}
  rpc CreateQueue(QueueInfo) returns  (google.protobuf.Empty) {}
  rpc DeleteQueue(QueueInfo) returns  (google.protobuf.Empty) {}
  rpc SendRequestApprovalResponse(RequestApprovalResponse) returns (google.protobuf.Empty) {}
  rpc SendRequestApprovalRequest(RequestApproval) returns (google.protobuf.Empty) {}
}

message InitRequest {
  string service_name = 1;
  string routing_key = 2;
  bool queue_auto_delete = 3;
}

message QueueInfo {
  string queue_name = 1;
  bool auto_delete = 2;
  string user_name = 3;
  string job_name = 4;
}

message ConsumeRequest {
  string queue_name = 1;
  bool auto_ack = 2;
}

message SideCarMessage {
  string type = 1;
  google.protobuf.Any body = 2;
  map<string, bytes> traces = 3;// Binary or textual representation of span context
}

message Auth {
  string access_token = 1;
  string refresh_token = 2;
}

message DataProvider {
  repeated string archetypes = 1;
  repeated string compute_providers = 2;
}

message ValidationResponse {
  string type = 1;
  string request_type = 2;
  map<string, DataProvider> valid_dataproviders = 3;
  repeated string invalid_dataproviders = 4;
  Auth auth = 5;
  User user = 6;
  bool request_approved = 7;
}

message User {
  string id = 1;
  string user_name = 2;
}

message RequestApproval {
  string type = 1;
  User user = 2;
  repeated string data_providers = 3;
  bool sync_services = 4;
  string destination_queue = 5;
}

message RequestApprovalResponse  {
  string type = 1;
  User user = 2;
  Auth auth = 3;
  map<string, string> authorized_providers = 4;
  string job_id = 5;
  string error = 6;
  RequestMetadata request_metadata = 7;
}

message PolicyUpdate {
  string type = 1;
  User user = 2;
  repeated string data_providers = 3;
  RequestMetadata request_metadata = 4;
  ValidationResponse validation_response = 5;
}

message CompositionRequest {
    string archetype_id = 1;
    string request_type = 2;
    string role = 3;
    User user = 4;
    repeated string data_providers = 5;
    string destination_queue = 6;
    string job_name = 7;
    string local_job_name = 8; // for saving the generated local job name with this composition request
}

message SqlDataRequest {
  string type = 1;
  string query = 2;
  bool graph = 3;
  string algorithm = 4;
  map<string, string> algorithm_columns = 5;
  User user = 6;
  RequestMetadata request_metadata = 7;
}
